### DISCUSSION TITLE
Combine switch expression with let keyword

### DISCUSSION THREAD

**Discussion by Khaos66** [dotnet/csharplang#4154] [Language Ideas] [answered]
It would be cool if it was possible to assign a new variable before the switch case expression is evaluated.
Something like this:
```c#
var result = value switch {
   string let var s = value.ToString() => s[..s.IndexOf('x')],
   _ => null
}
```

This is just a very simple example. Of course the let expression and case expressions can be more complex.
But this demonstrates, that _s_ is used multiple times inside the case expression and with a let expression _ToString()_ is only needed once.

Whats your oppinion on this?

**Comment by YairHalberstadt** [comment]
You can currently do:

```csharp
var result = value switch {
   string s => s[..s.IndexOf('x')],
   _ => null
}
```

**Reply by Khaos66** [reply]
Yes you're right. But as I sad the let could be something completely different. I just choose ToString, as it's simple to understand.
Here is a different example:
```c#
var result = type switch {
   Enum.TypeString let var s = value.ToString() => s[..s.IndexOf('x')],
   _ => null
}
```
This is actually more what I'm facing right now.

**Comment by Khaos66** [comment, marked as answer]
The best way to do this right now would be to call a method

```c#
string ConvertStr(string s) => s[..s.IndexOf('x')];

var result = type switch {
   Enum.TypeString => ConvertStr(value.ToString()),
   _ => null
}
```

So this is achievable right now. But I still want to leave the discussion open

**Reply by YairHalberstadt** [reply]
You can currently do this:

```csharp
var result = type switch {
   object o when o.ToString() is var s => s[..s.IndexOf('x')],
   _ => null
}
```

**Comment by stepanbenes** [comment]
> ```cs
> string let var s
> ```

This is brilliant.

**Comment by YairHalberstadt** [comment]
With https://github.com/dotnet/csharplang/issues/3037 you would be able to do:

```csharp
var result = type switch {
   Enum.TypeString value => {
       var s = value.ToString();
       s[..s.IndexOf('x')]
   },
   _ => null
}
```

**Reply by Khaos66** [reply]
Also acceptable ;)
